### DISCUSSION TITLE
How to introduce generic interface param to constructor

### DISCUSSION THREAD

**Discussion by paxhujing** [metalama/Metalama#52] [Q&A] [answered]
In asp.net core controller，inject ILogger<MyController> into constructors

**Comment by gfraiteur** [comment]
Could you elaborate the question with a code example please?

**Reply by paxhujing** [reply]
```
[ApiController]
[Route("[controller]/[action]")]
public class TestController : ControllerBase
{
    private readonly ILogger _logger;

    public TestController(ILogger<TestController> logger)
    {
        _logger = logger;
    }
}
```
Usually we need to inject ` ILogger<XXX>`
This pattern appears in many places, so i want to inject ` ILogger<XXX>` as an aspect.
```
internal class LoggerAttribute : ConstructorAspect
{
    public override void BuildAspect(IAspectBuilder<IConstructor> builder)
    {
        //How to introduce 'ILogger<XXX>'
        base.BuildAspect(builder);
    }
}

[ApiController]
[Route("[controller]/[action]")]
[Logger]
public class TestController : ControllerBase
{
}
```
This code should be equivalent to the above

**Comment by ghost** [comment]
Hello, 
This is a message to remind you that your request is pending and awaits your feedback.
Thank you
Zuzana

**Comment by gfraiteur** [comment]
The documentation for the case is here: https://doc.metalama.net/aspects/advising/introducing-constructor-parameters

**Comment by paxhujing** [comment, marked as answer]
```
public class LoggerAttribute : TypeAspect
{
    [Introduce(WhenExists = OverrideStrategy.Ignore)]
    private readonly ILogger _logger;

    public override void BuildAspect(IAspectBuilder<INamedType> builder)
    {
        Type type = typeof(ILogger<>).MakeGenericType(builder.Target.ToType());
        var constructors = builder.Target.Constructors;
        foreach (var constructor in constructors)
        {
            builder.Advice.IntroduceParameter(
                constructor,
                "logger",
                type,
                TypedConstant.Default(type),
                pullAction: (p, c) =>
                    PullAction.IntroduceParameterAndPull(
                        "logger",
                        TypeFactory.GetType(type),
                        TypedConstant.Default(type)
                        )
                );
            builder.Advice.AddInitializer(constructor,
                StatementFactory.Parse("_logger = logger;"));
        }
    }
}
```
```
[ApiController]
[Route("[controller]")]
[Logger]
public class WeatherForecastController : ControllerBase
{

}
```
I got a compilation error

> Error	LAMA0041	'LoggerAttribute.BuildAspect' threw 'NotSupportedException' when applied to 'WeatherForecastController': >Derived classes must provide an implementation. Exception details are in 'C:\Users\hj\AppData\Local\Temp\Metalama\CrashReports\0.5.51-preview\exception-f7902288-9791-40f0-8529-3843efc1370b.txt'. To >attach a debugger to the compiler, use the  '-p:MetalamaDebugCompiler=True' command-line option.	LobbyWebsite.Main	D:\Documents\JQK\A9\A9Lobby\website.src\dev\LobbyWebsite\Src\LobbyWebsite.Main\Controllers\WeatherForecastController.cs	6	

If do not use generic interfance `ILogger<T>`,it will compile successfully
```
    public override void BuildAspect(IAspectBuilder<INamedType> builder)
    {
        Type type = typeof(ILogger);
        //Type type = typeof(ILogger<>).MakeGenericType(builder.Target.ToType());
        ...
    }
```
and,there was another compilation error that confused me
```
    [Introduce(WhenExists = OverrideStrategy.Ignore)]
    private readonly ILogger _logger;
```
>Error	LAMA0611	Error CS0103 in code generated by aspect 'Logger': The name '_logger' does not exist in the current context	LobbyWebsite.Main	D:\Documents\JQK\A9\A9Lobby\website.src\dev\LobbyWebsite\Src\LobbyWebsite.Main\Controllers\WeatherForecastController.cs	42	

if i change 'private' to 'public',it will  compile successfully

**Reply by gfraiteur** [reply]
> if i change 'private' to 'public',it will compile successfully

There is a bug indeed. Template members must be public, but you can change their visibility using a property of the [Template] attribute
